ERROR_INVALID_IR (6) regression in numba-cuda 0.29 with CCCL transform iterators

## Summary

numba-cuda 0.29.0 (with numba 0.64.0) produces invalid NVVM IR when compiling CCCL's transform iterator examples and tests, causing `cuda.bindings.nvvm.nvvmError: ERROR_INVALID_IR (6)`. This is a regression — these tests pass with numba-cuda 0.26.x.

(Note: 0.27.x had a separate, unrelated `call_internal` breakage that was fixed in 0.28 via https://github.com/NVIDIA/numba-cuda/pull/782.)

## Failing tests

Consistent across all CI configs (GCC/MSVC, CTK 12.9/13.0/13.1, py3.10/py3.13, L4/H100):

- `test_compute_examples_iterator_transform_iterator_basic`
- `test_compute_examples_iterator_transform_iterator_lambda`
- `compute/test_iterators.py::test_transform_iterator_with_lambda`
- `compute/test_iterators.py::test_transform_iterator_with_zip_iterator`

## Traceback

```
_, return_type = numba.cuda.compile(...)
  File ".../numba_cuda/numba/cuda/core/compiler_lock.py", line 74, in _acquire_compile_lock
  File ".../numba_cuda/numba/cuda/compiler.py", line 1233, in compile
  File ".../numba_cuda/numba/cuda/codegen.py", line 235, in get_asm_strs
  File ".../numba_cuda/numba/cuda/cudadrv/nvvm.py", line 563, in compile_ir
  File ".../numba_cuda/numba/cuda/cudadrv/nvvm.py", line 236, in verify
cuda.bindings.nvvm.nvvmError: ERROR_INVALID_IR (6)
```

## Installed versions

- numba-cuda 0.29.0
- numba 0.64.0
- cuda-bindings 12.9.6

## Links

- CCCL PR where this was discovered: https://github.com/NVIDIA/cccl/pull/8180
- Example failing GHA job: https://github.com/NVIDIA/cccl/actions/runs/23571427217/job/68635892217

— Leo's bot

Reopening this issue because while I don't see the error from my [previous comment](https://github.com/NVIDIA/numba-cuda/issues/841#issuecomment-4169928693) anymore, I do still see the problem @leofang originally reported now, but without any further context:

```python

tests/compute/test_iterators.py:235:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
cuda/compute/iterators/_transform.py:82: in __init__
    value_type = self._transform_op.get_return_type(
cuda/compute/_jit.py:547: in get_return_type
    return _infer_return_type(self._func, input_types)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
cuda/compute/_caching.py:124: in inner
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
cuda/compute/_jit.py:456: in _infer_return_type
    _, return_type = numba.cuda.compile(
../../.venv/lib/python3.13/site-packages/numba_cuda/numba/cuda/core/compiler_lock.py:74: in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
../../.venv/lib/python3.13/site-packages/numba_cuda/numba/cuda/compiler.py:1245: in compile
    codes = lib.get_asm_strs(cc=cc)
            ^^^^^^^^^^^^^^^^^^^^^^^
../../.venv/lib/python3.13/site-packages/numba_cuda/numba/cuda/codegen.py:243: in get_asm_strs
    ptxes = [nvvm.compile_ir(irs, **options)]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../.venv/lib/python3.13/site-packages/numba_cuda/numba/cuda/cudadrv/nvvm.py:563: in compile_ir
    cu.verify()
../../.venv/lib/python3.13/site-packages/numba_cuda/numba/cuda/cudadrv/nvvm.py:236: in verify
    _nvvm.verify_program(
cuda/bindings/nvvm.pyx:215: in cuda.bindings.nvvm.verify_program
    ???
cuda/bindings/nvvm.pyx:235: in cuda.bindings.nvvm.verify_program
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   cuda.bindings.nvvm.nvvmError: ERROR_INVALID_IR (6)

cuda/bindings/nvvm.pyx:58: nvvmError
============================================================================================= short test summary info ==============================================================================================
FAILED tests/compute/test_iterators.py::test_transform_iterator_with_zip_iterator - cuda.bindings.nvvm.nvvmError: ERROR_INVALID_IR (6)
```

Working on a minimal repro.

--- a/numba_cuda/numba/cuda/core/imputils.py
+++ b/numba_cuda/numba/cuda/core/imputils.py
@@ -222,7 +222,9 @@ def user_function(fndesc, libs):
 
         if status is not None:
             with cgutils.if_unlikely(builder, status.is_error):
-                fndesc.call_conv.return_status_propagate(builder, status)
+                context.fndesc.call_conv.return_status_propagate(
+                    builder, status
+                )
 
         assert sig.return_type == fndesc.restype
 

--- a/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
+++ b/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
@@ -837,7 +837,7 @@ class TestCompile(unittest.TestCase):
             wrapper_func, types.int32(types.int32), output="ltoir", abi="c"
         )
 
-    def test_compile_CABI_calling_overloaded_function(self):
+    def test_compile_cabi_calling_overloaded_function(self):
         # Reproducer from gh-841
         # https://github.com/NVIDIA/numba-cuda/issues/841
         #

--- a/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
+++ b/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
@@ -817,7 +817,7 @@ class TestCompile(unittest.TestCase):
 
         cuda.compile(wrapper, wrapper_sig.args, output="ltoir")
 
-    def test_compile_CABI_calling_device_function_returning_optional(self):
+    def test_compile_cabi_calling_device_function_returning_optional(self):
         # Exercise a CABI caller invoking a Numba ABI callee that can return
         # None through Optional[int32]
         def maybe_none(x):

--- a/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
+++ b/numba_cuda/numba/cuda/tests/cudapy/test_compiler.py
@@ -4,7 +4,7 @@
 import os
 from math import sqrt
 from numba import cuda
-from numba.core.extending import intrinsic
+from numba.core.extending import intrinsic, overload
 
 from numba.cuda import float32, int16, int32, int64, types, uint32, void
 from numba.cuda import (
@@ -836,6 +836,34 @@ class TestCompile(unittest.TestCase):
         cuda.compile(
             wrapper_func, types.int32(types.int32), output="ltoir", abi="c"
         )
+
+    def test_compile_CABI_calling_overloaded_function(self):
+        # Reproducer from gh-841
+        # https://github.com/NVIDIA/numba-cuda/issues/841
+        #
+        # When a CABI function calls an overloaded function (compiled as a
+        # Numba-ABI subroutine), the error-status return path must use the
+        # caller's calling convention. Previously the callee's Numba-ABI
+        # convention was used, emitting `ret i32` (the status code) inside a
+        # function whose declared return type did not match, producing invalid
+        # LLVM IR.
+
+        def _my_func(x):
+            pass
+
+        @overload(_my_func)
+        def _ov_my_func(x):
+            if isinstance(x, types.IntegerLiteral):
+
+                def impl(x):
+                    return int32(0)
+
+                return impl
+
+        def caller():
+            _my_func(0)
+
+        cuda.compile(caller, (), abi_info={"abi_name": "caller"}, abi="c")
 
     def test_compile_complex_div_c_abi(self):
         # Reproducer from gh-789

--- a/numba_cuda/numba/cuda/tests/cudapy/test_device_func.py
+++ b/numba_cuda/numba/cuda/tests/cudapy/test_device_func.py
@@ -557,6 +557,34 @@ class TestDeclareDeviceCABI(CUDATestCase):
         kernel[1, 32](r, x)
         np.testing.assert_equal(r, x * 2)
 
+    def test_declare_device_cabi_propagate(self):
+        # Ensures that a device function calling a C ABI function correctly
+        # propoagates the error status to its caller. No error status is
+        # provided by the C ABI function, so the device function that calls it
+        # should return status 0 (no error).
+        times2 = cuda.declare_device(
+            "times2", "int32(int32)", link=times2_cabi_cu, abi="c"
+        )
+
+        @cuda.jit
+        def device_func(x):
+            return times2(x)
+
+        @cuda.jit
+        def kernel(r, x):
+            i = cuda.grid(1)
+            if i < len(r):
+                r[i] = device_func(x[i])
+
+        x = np.arange(10, dtype=np.int32)
+        r = np.empty_like(x)
+        kernel[1, 32](r, x)
+        np.testing.assert_equal(r, x * 2)
+
+        compile_result = device_func.overloads[device_func.signatures[0]]
+        device_func_ir = compile_result.library.get_llvm_str()
+        self.assertIn("ret i32 0", device_func_ir)
+
     def test_declare_device_cabi_zero_args(self):
         const42 = cuda.declare_device(
             "const42", "int32()", link=const42_cabi_cu, abi="c"

